Streamed sources implementation.

ibuenros opened the pull request.

ibuenros commented:
@ydai1124 can you review?

shirshanka commented on gobblin-api/src/main/java/gobblin/source/StreamedSource.java:
```diff
@@ -0,0 +1,39 @@
+ *    processing available work units while future work units are still being computed.
+ *  * Infinite work unit streams: in the future, some job launchers will support infinite streams of work units.
+ */
+public interface StreamedSource<S, D> extends Source<S, D> {
```
This is an overloaded and confusing term... need to think of a better name... source that produces stream of work-units versus a source of streaming data...

ibuenros commented:
@shirshanka addressed the comments around using object attributes instead of instance checks. Also, do you have a suggestion on a better name for `StreamedSource`?

shirshanka commented on gobblin-api/src/main/java/gobblin/source/workunit/WorkUnitStream.java:
```diff
@@ -0,0 +1,144 @@
+ * A stream of {@link WorkUnit}s, allows for working with large numbers of work units in a memory-efficient way, as
+ * well as processing infinite streams of work units.
+ */
+public class WorkUnitStream {
```
Given that this will be a pretty important Type for all Sources to implement (eventually), should it be an interface?

ibuenros commented:
@shirshanka made the `WorkUnitStream` an interface.

shirshanka commented on gobblin-runtime/src/test/java/gobblin/runtime/TestStreamedSource.java:
```diff
@@ -0,0 +1,191 @@
+
+
+@Slf4j
+public class TestStreamedSource {
```
For consistency, this should be TestWorkUnitStreamSource
